feat: add XDeFi Adapter

loxator opened the pull request:
Addition of XDeFi Adapter to work with the XDeFi Wallet. Currently, only supporting the Ethereum chain, and will be used to add more chains further.

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  public ethNextAccountPath(msg: core.ETHAccountPath): core.ETHAccountPath | undefined {
+    // TODO: What do we do here?
```
This method isn't implemented in anything else, and it's really only in hdwallet for legacy reasons; account paths are (will be?) taken care of by chain-adapters now. (We need to do an API scrub now that we've dropped support for the old beta.shapeshift.com stuff.)

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+    return "XDeFi";
+  }
+
+  public hasOnDevicePinEntry(): boolean {
+    return false;
+  }
+
+  public hasOnDevicePassphrase(): boolean {
+    return true;
+  }
+
+  public hasOnDeviceDisplay(): boolean {
+    return true;
+  }
+
+  public hasOnDeviceRecovery(): boolean {
+    return true;
+  }
+
+  public hasNativeShapeShift(srcCoin: core.Coin, dstCoin: core.Coin): boolean {
+    // It doesn't... yet?
+    return false;
+  }
```
This is all legacy BS and will be burned with fire at some point.

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+    return false;
+  }
+
+  public ethSupportsNativeShapeShift(): boolean {
+    return false;
+  }
```
(fyi, legacy BS again)

reidrankin commented on packages/hdwallet-xdefi/package.json:
```diff
@@ -0,0 +1,29 @@
+    "prepublishOnly": "yarn clean && yarn build"
+  },
+  "dependencies": {
+    "@portis/web3": "3.0.10",
```
betting some of these deps need cleanup

reidrankin commented on packages/hdwallet-xdefi/src/ethereum.ts:
```diff
@@ -0,0 +1,125 @@
+import * as core from "@shapeshiftoss/hdwallet-core";
+import { ETHSignedMessage } from "@shapeshiftoss/hdwallet-core";
+
+export function describeETHPath(path: core.BIP32Path): core.PathDescription {
+  let pathStr = core.addressNListToBIP32(path);
+  let unknown: core.PathDescription = {
+    verbose: pathStr,
+    coin: "Ethereum",
+    isKnown: false,
+  };
+
+  if (path.length !== 5) return unknown;
+
+  if (path[0] !== 0x80000000 + 44) return unknown;
+
+  if (path[1] !== 0x80000000 + core.slip44ByCoin("Ethereum")) return unknown;
+
+  if ((path[2] & 0x80000000) >>> 0 !== 0x80000000) return unknown;
+
+  if (path[3] !== 0) return unknown;
+
+  if (path[4] !== 0) return unknown;
+
+  let index = path[2] & 0x7fffffff;
+  return {
+    verbose: `Ethereum Account #${index}`,
+    accountIdx: index,
+    wholeAccount: true,
+    coin: "Ethereum",
+    isKnown: true,
+  };
+}
```
You can get this logic from `core.describeETHPath()`

reidrankin commented on packages/hdwallet-xdefi/src/ethereum.ts:
```diff
@@ -0,0 +1,125 @@
+  };
+}
+
+export async function ethVerifyMessage(msg: core.ETHVerifyMessage, ethereum: any): Promise<boolean | null> {
+  console.error("Method ethVerifyMessage unsupported for XDeFi wallet!");
+  return null;
+}
```
This can be polyfilled in if we can get the public key out of xdefi

reidrankin replied:
Actually, now that I think about it, we don't need the public key -- we can just recover it from the message signature, calculated the address from the key, and compare it with what it's supposed to be

reidrankin commented on packages/hdwallet-xdefi/src/adapter.ts:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  public async pairDevice(): Promise<core.HDWallet> {
+    const provider: any = (window as any).xfi && (window as any).xfi.ethereum;
+    if (!provider) {
+      window.open("https://xdefi.io", "_blank");
+    }
```
It's not appropriate to fire window.open here; better to throw and let the application handle it. My instinct is also to use `globalThis` instead of `window` but idk.

loxator replied:
You have a very good point, I will throw an exception here and handle it on the client side. Will also check between `this` vs `window`.

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+    return "XDeFi";
+  }
+
+  public getModel(): Promise<string> {
+    return Promise.resolve("XDeFi");
+  }
```
```suggestion
  public async getModel(): Promise<string> {
    return "XDeFi";
  }
```

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+    return Promise.resolve("XDeFi");
+  }
+
+  public getLabel(): Promise<string> {
+    return Promise.resolve("XDeFi");
+  }
```
```suggestion
  public async getLabel(): Promise<string> {
    return "XDeFi";
  }
```

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+    return Promise.resolve("XDeFi");
+  }
+
+  public async initialize(): Promise<any> {
+    try {
+      this.provider = (window as any).xfi && (window as any).xfi.ethereum;
+    } catch (e) {
+      console.error(e);
+    }
+
+    return Promise.resolve();
+  }
```
returning `any` kills kittens
```suggestion
  public async initialize(): Promise<void> {
    try {
      this.provider = (window as any).xfi && (window as any).xfi.ethereum;
    } catch (e) {
      console.error(e);
    }
  }
```

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  public loadDevice(msg: core.LoadDevice): Promise<void> {
+    // TODO: Does XDeFi allow this to be done programatically?
```
(probably not, and that's fine)

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+type HasNonTrivialConstructor<T> = T extends { new (): any } ? never : T;
+
+export class XDeFiHDWallet implements core.HDWallet, core.ETHWallet {
+  readonly _supportsETH = true;
+  readonly _supportsETHInfo = true;
+  readonly _supportsBTCInfo = false;
+  readonly _supportsBTC = false;
+  readonly _supportsCosmosInfo = false;
+  readonly _supportsCosmos = false;
+  readonly _supportsOsmosisInfo = false;
+  readonly _supportsOsmosis = false;
+  readonly _supportsBinanceInfo = false;
+  readonly _supportsBinance = false;
+  readonly _supportsDebugLink = false;
+  readonly _isPortis = false;
+  readonly _isMetaMask = false;
+  readonly __XDEFI = true;
+  readonly _supportsRippleInfo = false;
+  readonly _supportsRipple = false;
+  readonly _supportsEosInfo = false;
+  readonly _supportsEos = false;
+  readonly _supportsFioInfo = false;
+  readonly _supportsFio = false;
+  readonly _supportsThorchainInfo = false;
+  readonly _supportsThorchain = false;
+  readonly _supportsSecretInfo = false;
+  readonly _supportsSecret = false;
+  readonly _supportsKava = false;
+  readonly _supportsKavaInfo = false;
+  readonly _supportsTerra = false;
+  readonly _supportsTerraInfo = false;
```
- nothing good will come of breaking the naming convention
> _alles muss in ordnung kommen_
- all the `= false` can die
```suggestion
  readonly _supportsETH = true;
  readonly _isXDeFi = true;
```

reidrankin reviewed (changes requested):
First off, thanks for this PR! We appreciate it greatly and look forward to getting it in. I've taken a first-pass at this and left some suggestions; it may be easiest to drop into the Engineering section of our Discord server and workshop this live.

Also, one thing that we will need before merge is a comprehensive set of unit and integration tests, which will probably require cooking up some sort of xdefi mock if they don't have one already.

loxator commented:
Hi @reidrankin,

Thanks for the awesome review and raising some very valid points. I am resolving them one by one as I go, some are pretty straightforward, but I would like a bit of clarification on a couple of them.

The Adapter which was committed is a copy of your MetaMask adapter. The idea is to start of XDeFi with the EVM based chains, and then move onto other chains.

You have marked some of the methods as `obsolete` but they still exist in the interface that the adapter implements. For example, some of the methods come from the interface `core.HDWallet`, since you mentioned that they are going to be removed at some point, should I leave the methods in for now, and can be refactored later on when the interface changes? Or should I also remove them from both the places in this PR?

Also, it'd be great if there is an example I can look at for implementing the unit/integration tests so I conform to your repository's standards.

Thanks a lot again for taking the time to review and going over the PR, I am still new to the world of wallets and learning as I go along, so I will have a lot of questions going forward :sweat_smile: !

reidrankin commented:
Sorry for the confusion about the obsolescence of various things; they do still have to be implemented when required by the interfaces, but hopefully knowing our plans to nuke them will give you some useful context.

For test references, check out `integration/src/*.test.ts`, `integration/src/wallets/*.ts`. You'll need to:
- implement any appropriate unit tests for your new package; our best-tested packages are `hdwallet-native` and `hdwallet-keepkey`, so take a look at `packages/hdwallet-{native,keepkey}/src/*.test.ts` to get an idea of what's needed
- cut-n-paste `integration/src/foobar.ts` to `integration/src/xdefi.ts`
- implement the `WalletSuite` interface from `integration/src/wallets/suite.ts` in `integration/src/wallets/xdefi.ts`; I'd start by copying the portis one, since that's a web3 sort of thing too
-  you'll probably need to mock the parts of xdefi you need to make your code testable; tests must be consistently reproducible, which means they must pass without a network connection

As it turns out, `hdwallet-metamask` lacks these sorts of tests; I remember it was implemented under a certain amount of time pressure, which is probably why we dropped the ball. However, I have a feeling that making the tests needed for `hdwallet-xdefi` and `hdwallet-metamask` will be highly similar, and I have another feeling that our engineering workstream will be more than willing to throw some dosh in your general direction for eliminating this particular oversight. (@0xdef1cafe back me up on this?)

reidrankin commented on packages/hdwallet-xdefi/src/adapter.ts:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  public async pairDevice(): Promise<core.HDWallet> {
+    const provider: any = (window as any).xfi && (window as any).xfi.ethereum;
```
```suggestion
    const provider: any = (globalThis as any).xfi?.ethereum;
```

reidrankin commented on packages/hdwallet-xdefi/src/adapter.ts:
```diff
@@ -0,0 +1,42 @@
+      console.error("Could not get XDeFi accounts. ");
+      throw error;
+    }
+    const wallet = new XDeFiHDWallet();
+    await wallet.initialize();
```
Just noticed that `provider` doesn't actually get passed into the instance here!

loxator replied:
The `provider` is not passed as that's how it the method signature is defined in `code.HDWallet`. Should the provider be passed into the instance instead?

```
  /**
   * Initialize a device session.
   */
  initialize(): Promise<any>;
  ```

mrnerdhair replied:
Ah, gotcha. We can still extend that interface, we just have to do it a bit wonkily -- see the matching comment on the wallet implementation for details.
```suggestion
    const wallet = new XDeFiHDWallet();
    await wallet.initialize(provider);
```

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+}
+
+export class XDeFiHDWalletInfo implements core.HDWalletInfo, core.ETHWalletInfo {
+  readonly _supportsBTCInfo = false;
+  readonly _supportsETHInfo = true;
+  readonly _supportsCosmosInfo = false;
+  readonly _supportsBinanceInfo = false;
+  readonly _supportsRippleInfo = false;
+  readonly _supportsEosInfo = false;
+  readonly _supportsFioInfo = false;
+  readonly _supportsThorchainInfo = false;
+  readonly _supportsSecretInfo = false;
+  readonly _supportsKavaInfo = false;
+  readonly _supportsTerraInfo = false;
```
```suggestion
  readonly _supportsETHInfo = true;
```

reidrankin commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,340 @@
+    return true;
+  }
+
+  public disconnect(): Promise<void> {
+    return Promise.resolve();
+  }
```
```suggestion
  public async disconnect(): Promise<void> { }
```

loxator replied:
This needs a promise to be resolved as that's how the signature is in the interface `core.HDWallet`, otherwise Typescript complains.

cjthompson replied:
`async` guarantees a `Promise` is returned so I would think this should be fine.

loxator replied:
That is true, but we aren't returning anything in the suggestion right? Typescript complains that a value must be returned 
`A function whose declared type is neither 'void' nor 'any' must return a value.ts(2355)`

loxator replied:
Please ignore the previous comment as I missed out on the `async` bit.

loxator commented:
@reidrankin 

I have tried to add unit and integration tests according to your instructions.

I was able to get 100% coverage for most of the unit test files, except `xdefi.ts`. I have left this one for now as a lot of methods are just Promises being resolved due to the implemented interface for `core.HDWallet`.

I have tried to set up the integration tests as you mentioned, but have a bit of difficulty understanding  on how to change some tests that`suite.ts` runs. For example, our wallet doesn't support verification of signed messages, yet there is a test in `ethereumTests` that `suites.ts` runs on `xdefi.ts` in the integration tests, how do I overwrite the necessary tests?

Lastly, thanks a lot for your help!

loxator commented:
@reidrankin,

I have added the necessary tests and also implemented the missing functions for `ethVerifyMessage` and `ethSignTransaction`. 

Please let me know if anything else is required from my end? Once this adapter is merged, then I can proceed with the Bitcoin Adapter. If changes are required, please let me know soon as the XDeFi team would like this to be included as soon as possible. 

Thanks a lot for all your help!

loxator commented:
@reidrankin @pastaghost 

Just bumping this PR up as we would like to add the integration for EVM chains soon. Thank you :pray: !

loxator commented:
@cjthompson I have updated the PR according to your comments.

cjthompson commented:
@mrnerdhair - the build is failing because it can't find the `@shapeshiftoss/hdwallet-xdefi` package, which obviously isn't published. I would think that it would be better for the build to use the local code in `packages` vs downloading published packages. Can we look at fixing module resolution to resolve locally vs. pulling from npm?

mrnerdhair commented:
@loxator I think I've fixed the build issue here. I've PRed the fix into this branch: https://github.com/XDeFi-tech/hdwallet/pull/1

mrnerdhair commented on packages/hdwallet-xdefi/src/index.test.ts:
```diff
@@ -0,0 +1,11 @@
+import * as library from "./";
+
+describe("Exports all expected classes", () => {
+  it("should export EncryptedWallet", () => {
```
```suggestion
  it("should export XDeFiAdapter", () => {
```

mrnerdhair commented on packages/hdwallet-xdefi/src/index.test.ts:
```diff
@@ -0,0 +1,11 @@
+    expect(library.XDeFiAdapter.name).toBe("XDeFiAdapter");
+  });
+
+  it("should export WebCryptoEngine", () => {
```
```suggestion
  it("should export XDeFiHDWallet", () => {
```

mrnerdhair commented on packages/hdwallet-xdefi/src/xdefi.ts:
```diff
@@ -0,0 +1,288 @@
+    return "XDeFi";
+  }
+
+  public async initialize(): Promise<any> {
+    try {
+      this.provider = (window as any).xfi && (window as any).xfi.ethereum;
+    } catch (e) {
+      console.error(e);
+    }
+  }
```
The first overload provides a static hint that the method will throw without options given. The implementation signature complies with the core interface but isn't visible itself from other code, avoiding any ambiguity about `undefined` being a supported value.

```suggestion
  public initialize(): never;
  public initialize(provider: unknown): Promise<any>;
  public async intialize(provider?: unknown): Promise<any> {
    if (!provider) throw new Error("provider is required");
    this.provider = provider;
  }
```

mrnerdhair reviewed (approved):
Excellent work here. I've PRed the build fix as XDeFi-tech#1; once that is integrated and these comments are resolved, LGTM.

mrnerdhair commented:
Migrated to #405.
